src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,155 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node<T> head;
+    private Node<T> tail;
+
+    private class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> addedNode = new Node<>(tail, value, null);
+        if (isEmpty()) {
+            head = addedNode;
+        } else {
+            tail.next = addedNode;
+        }
+        tail = addedNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Can't ADD element by that index");
+        } else if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> addingNode = new Node<T>(null, value, head);
+            head = addingNode;
+            size++;
+        } else {
+            Node<T> previousNode = findNodeByIndex(index).prev;
+            Node<T> nextNode = previousNode.next;
+            Node<T> addingNode = new Node<>(previousNode, value, nextNode);
+            nextNode.prev = addingNode;
+            previousNode.next = addingNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T addedNode : list) {
+            add(addedNode);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, "get");
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, "set");
+        Node<T> nodeToSet = findNodeByIndex(index);
+        T oldValue = nodeToSet.value;
+        nodeToSet.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, "remove");
+        if (isEmpty()) {
+            return null;
+        }
```

Comment:
do we meed this check here?

Reply:
Seems nope cause of findNodeByIndex(index) may return null in some cases